pkg/apis/eventing/v1alpha1/broker_types.go

```diff
@@ -0,0 +1,156 @@
+var _ apis.Validatable = (*Broker)(nil)
+var _ apis.Defaultable = (*Broker)(nil)
+var _ apis.Immutable = (*Broker)(nil)
+var _ runtime.Object = (*Broker)(nil)
+var _ webhook.GenericCRD = (*Broker)(nil)
+
+type BrokerSpec struct {
+	// TODO By enabling the status subresource metadata.generation should increment
+	// thus making this property obsolete.
+	//
+	// We should be able to drop this property with a CRD conversion webhook
+	// in the future
+	//
+	// +optional
+	DeprecatedGeneration int64 `json:"generation,omitempty"`
+
+	ChannelTemplate *ChannelSpec `json:"channelTemplate,omitempty"`
+}
+
+var brokerCondSet = duckv1alpha1.NewLivingConditionSet(BrokerConditionIngress, BrokerConditionChannel, BrokerConditionFilter, BrokerConditionAddressable)
+
+// BrokerStatus represents the current state of a Broker.
+type BrokerStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Broker.
+	// It corresponds to the Broker's generation, which is updated on mutation by
+	// the API Server.
+	// TODO: The above comment is only true once
+	// https://github.com/kubernetes/kubernetes/issues/58778 is fixed.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the latest available observations of a broker's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions duckv1alpha1.Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// Broker is Addressable. It currently exposes the endpoint as a
+	// fully-qualified DNS name which will distribute traffic over the
+	// provided targets from inside the cluster.
+	//
+	// It generally has the form {broker}-router.{namespace}.svc.{cluster domain name}
+	Address duckv1alpha1.Addressable `json:"address,omitempty"`
+}
+
+const (
+	BrokerConditionReady = duckv1alpha1.ConditionReady
+
+	BrokerConditionIngress duckv1alpha1.ConditionType = "Ingress"
+
+	BrokerConditionChannel duckv1alpha1.ConditionType = "Channel"
+
+	BrokerConditionFilter duckv1alpha1.ConditionType = "Filter"
+
+	BrokerConditionAddressable duckv1alpha1.ConditionType = "Addressable"
+)
+
+// GetCondition returns the condition currently associated with the given type, or nil.
+func (bs *BrokerStatus) GetCondition(t duckv1alpha1.ConditionType) *duckv1alpha1.Condition {
+	return brokerCondSet.Manage(bs).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (bs *BrokerStatus) IsReady() bool {
+	return brokerCondSet.Manage(bs).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (bs *BrokerStatus) InitializeConditions() {
+	brokerCondSet.Manage(bs).InitializeConditions()
+}
+
+func (bs *BrokerStatus) MarkIngressReady() {
+	brokerCondSet.Manage(bs).MarkTrue(BrokerConditionIngress)
+}
+
+func (bs *BrokerStatus) MarkIngressFailed(err error) {
+	brokerCondSet.Manage(bs).MarkFalse(BrokerConditionIngress, "failed", "%v", err)
```

Comment:
failed is not a very good reason. I would pass in the reason in to the function.

Reply:
err has all the real information. It could have been caused by numerous problems (e.g. error getting the deployment, error creating the deployment, etc.) and I don't really want to thread them all the way through. The error message should be sufficient. If we find it isn't, then we can thread the reasons through the controller logic.

Reply:
I would agree with Scott here. It could be one word but slightly more descriptive then failed. This "reason" will be copied over as Ready condition's reason. As per the yaml files, when we anyone does kubectl get borkers mybroker. They will see something like:
Name:MyBroker, READY:False, REASON:failed, which doesn't help much.
 The only way to get the logged message is by running describe on it. 
Same comment in other places too including trigger
reference from yaml file
    - name: Ready
      type: string
      JSONPath: ".status.conditions[?(@.type==\"Ready\")].status"
    - name: Reason
      type: string
      JSONPath: ".status.conditions[?(@.type==\"Ready\")].reason"

Reply:
I agree that this should be better across the board. 

However, none of current code does any better. Looking at `Channel`'s `MarkNotProvisioned`, all callers have a reason hard coded to `NotProvisioned`. Similarly, there is only one caller of `Channel`'s `MarkProvisionerNotInstalled` and it hardcodes "Provisioner not found." (as an aside, this violates [K8s norms](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) "one-word CamelCase reason for the condition's last transition").

Reply:
#895, which also lead to #896.